src/core/data/grid/grid.hpp

```diff
@@ -0,0 +1,97 @@
+#ifndef PHARE_CORE_DATA_GRID_GRID_BASE_HPP
+#define PHARE_CORE_DATA_GRID_GRID_BASE_HPP
+
+#include <array>
+#include <cstddef>
+#include <cassert>
+#include <string>
+#include <utility>
+#include <vector>
+#include <algorithm>
+
+#include "core/def.hpp"
+#include "core/data/field/field.hpp"
+#include "core/data/ndarray/ndarray_vector.hpp"
+
+namespace PHARE::core
+{
+
+
+template<typename NdArrayImpl, typename PhysicalQuantity>
+class Grid : public NdArrayImpl
+{
+    using Super = NdArrayImpl;
+
+public:
+    using value_type             = typename NdArrayImpl::type;
+    using physical_quantity_type = PhysicalQuantity;
+    using NdArrayImpl::dimension;
+    using field_type = Field<NdArrayView<dimension, value_type>, PhysicalQuantity>;
+
+
+    Grid()                              = delete;
+    Grid(Grid const& source)            = delete;
+    Grid(Grid&& source)                 = default;
+    Grid& operator=(Grid&& source)      = delete;
+    Grid& operator=(Grid const& source) = delete;
+
+    template<typename... Dims>
+    Grid(std::string const& name, PhysicalQuantity qty, Dims... dims)
+        : Super{dims...}
+        , name_{name}
+        , qty_{qty}
+        , field_{name, qty, Super::data(), Super::shape()}
+    {
+        static_assert(sizeof...(Dims) == dimension, "Invalid dimension");
+    }
+
+    template<std::size_t dim>
+    Grid(std::string const& name, PhysicalQuantity qty, std::array<std::uint32_t, dim> const& dims)
+        : Super{dims}
+        , name_{name}
+        , qty_{qty}
+        , field_{name, qty, Super::data(), Super::shape()}
+    {
+    }
+
+    NO_DISCARD std::string name() const { return name_; }
+
+    NO_DISCARD constexpr PhysicalQuantity physicalQuantity() const { return qty_; }
+
+    template<typename That>
+    void copyData(That const& that)
+    {
+        assert(Super::size() == that.size());
```

Comment:
I dunno, depends on teh client case, is it ever the case where the shape can change and the size doesn't?
I don't think there is, so it would make sense for this when the Qty doesn't change